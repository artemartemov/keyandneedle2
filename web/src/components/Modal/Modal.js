import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import X from 'react-feather/dist/icons/x';
import { Transition } from 'react-transition-group';
import { colors, mq } from 'utils';

import Portal from '../../hooks/createPortal';

const AbsoluteCenter = css`
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  max-height: calc(100% - 100px);
`;

export const ModalWrapper = styled.div`
  width: 100vw;
  max-width: ${props => props.maxModalWidth};
  background: ${props => props.background};
  max-height: ${props => props.maxHeight};
  overflow: ${props => props.overflow || 'scroll'};
  z-index: 999999;

  ${mq.desktop`
    width: ${props => props.modalWidth};
  `}

  ${props => (props.centered ? AbsoluteCenter : null)};

  ${props =>
    props.issidepanel
      ? css`
          position: fixed;
          top: 0;
        `
      : null};

  ${props =>
    props.issidepanelright
      ? css`
          position: fixed;
          top: 0;
          right: 0;
          bottom: 0;
        `
      : null};
`;

const ModalHeader = styled.div`
  display: flex;
  font-size: 0.6875rem;
  line-height: 1.2;
  letter-spacing: 0.1rem;
  text-transform: uppercase;
  font-weight: bold;
  color: white;
  height: 3rem;
  position: fixed;
  width: 100%;
  justify-content: flex-end;
`;

const ModalTitle = styled.h2`
  width: 100%;
  margin: 0;
  font-weight: normal;
  font-size: 0.875rem;
  line-height: 3.5;
  color: white;
  letter-spacing: 0.1rem;
  text-transform: uppercase;
  text-align: center;
`;

const IconContainer = styled.div`
  width: 3rem;
  height: 3rem;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 0.25rem;
  cursor: pointer;
  float: right;
`;

const ModalBackground = styled.span`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: block;
  background: #0000009c;
  opacity: 0;
  transition: opacity 1s linear;

  ${props => (props.isActive ? 'opacity: 1;' : null)}
`;

const CloseIcon = styled(X).attrs({
  size: 32,
})``;

const Modal = ({
  issidepanel,
  issidepanelright,
  centered,
  background,
  modalWidth,
  maxModalWidth,
  onClose,
  title,
  children,
  className,
  onKeyDown,
  isModalOpen,
  onEnter,
  onExit,
  timeout,
}) => {
  const closeIconWrapper = (
    <IconContainer onClick={onClose} data-testid="modal-close">
      <CloseIcon color={issidepanel ? colors.beige.light : colors.grey.dark} />
    </IconContainer>
  );

  const modalMarkup = (
    <>
      <Transition in={isModalOpen} onEnter={onEnter} onExit={onExit} unmountOnExit timeout={timeout}>
        <Portal>
          <ModalWrapper
            issidepanel={issidepanel}
            issidepanelright={issidepanelright}
            className={className}
            modalWidth={modalWidth}
            maxModalWidth={maxModalWidth}
            background={background}
            centered={centered}
            onKeyDown={onKeyDown}
            data-testid="modal-wrapper"
          >
            <ModalHeader>
              {(issidepanel || issidepanelright) && closeIconWrapper}
              {issidepanel ||
                (issidepanelright !== true && title && <ModalTitle data-testid="modal-header">{title}</ModalTitle>)}
            </ModalHeader>
            {children}
          </ModalWrapper>
        </Portal>
      </Transition>

      {isModalOpen && <ModalBackground onClick={onClose} isActive={isModalOpen} />}
    </>
  );

  return modalMarkup;
};

Modal.propTypes = {
  /**
   * Child nodes
   */
  children: PropTypes.node.isRequired,
  /**
   * Close handler for modal button
   */
  onClose: PropTypes.func,
  /**
   * Title text to show in header
   */
  title: PropTypes.string.isRequired,
  /**
   * Whether to show a close icon
   */
  showCloseIcon: PropTypes.bool,
};

Modal.defaultProps = {
  onClose: null,
  showCloseIcon: true,
};

export default Modal;
