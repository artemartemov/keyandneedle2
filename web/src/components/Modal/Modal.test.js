import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { toHaveTextContent } from '@testing-library/jest-dom';
import 'jest-styled-components';
import Modal, { ModalWrapper } from './Modal';

afterEach(cleanup);

expect.extend({ toHaveTextContent });

const modal = <Modal title="Who is a modal?">I am a modal</Modal>;

describe('<Modal />', () => {
  it('#render() ModalWrapper centered', () => {
    const { container } = render(<ModalWrapper centered />);
    expect(container.firstChild).toHaveStyleRule('transform', 'translate(-50%,-50%)');
    expect(container.firstChild).toHaveStyleRule('top', '50%');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('#render() is it a sidepanel?', () => {
    const { baseElement, getByTestId } = render(
      <Modal title="Who is a modal?" isSidePanel>
        I am a modal
      </Modal>
    );
    expect(getByTestId('modal-wrapper')).toHaveStyleRule('position', 'fixed');
    expect(getByTestId('modal-wrapper')).toHaveStyleRule('top', '0');
    expect(baseElement).toMatchSnapshot();
  });

  it('#render() default without props defined', () => {
    const { baseElement, getByTestId } = render(modal);
    expect(getByTestId('modal-header')).toHaveTextContent('Who is a modal?');
    expect(baseElement).toHaveTextContent('I am a modal');
    expect(baseElement).toMatchSnapshot();
  });
});
