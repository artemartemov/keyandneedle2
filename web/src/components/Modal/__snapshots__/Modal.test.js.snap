// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> #render() ModalWrapper centered 1`] = `
.c0 {
  overflow: scroll;
  z-index: 999999;
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  max-height: calc(100% - 100px);
}

<div
  class="c0"
/>
`;

exports[`<Modal /> #render() default without props defined 1`] = `
<body>
  <div />
  .c0 {
  overflow: scroll;
  z-index: 999999;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.6875rem;
  line-height: 1.2;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  text-transform: uppercase;
  font-weight: bold;
  color: white;
  height: 3rem;
  position: fixed;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c2 {
  width: 100%;
  margin: 0;
  font-weight: normal;
  font-size: 0.875rem;
  line-height: 3.5;
  color: white;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  text-transform: uppercase;
  text-align: center;
}

<div
    class="c0"
    data-testid="modal-wrapper"
  >
    <div
      class="c1"
    >
      <h2
        class="c2"
        data-testid="modal-header"
      >
        Who is a modal?
      </h2>
    </div>
    I am a modal
  </div>
</body>
`;

exports[`<Modal /> #render() is it a sidepanel? 1`] = `
.c0 {
  overflow: scroll;
  z-index: 999999;
  position: fixed;
  top: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.6875rem;
  line-height: 1.2;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  text-transform: uppercase;
  font-weight: bold;
  color: white;
  height: 3rem;
  position: fixed;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c2 {
  width: 3rem;
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 0.25rem;
  cursor: pointer;
  float: right;
}

<body>
  <div />
  <div
    class="c0"
    data-testid="modal-wrapper"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        data-testid="modal-close"
      >
        <svg
          fill="none"
          height="32"
          stroke="#ebcb80"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </div>
    </div>
    I am a modal
  </div>
</body>
`;
